### PR TITLE
Sentry integration

### LINKED ISSUE #90: Integrate Sentry.io
[Sentry](https://sentry.io) is a service that provides intelligence into exceptions that occur within the application. They have generously sponsored us with an open source license so it's time we start integrating it.

The following issues need to be considered when integrating:

* The integration will need to be disabled in debug builds, otherwise the console will be spammed with useless reports. Probably can use the same check that disables the update checker
* Need to check what information is included in the reports. May need to exclude some information due to privacy concerns
* Need to provide to disable the reporting in case the player doesn't want to share information about their installation
* Need to decide on a way to ID the users. Possible candidates: steam id, machine ID

### PR DESCRIPTION
**Pull request by RevZero**
Integrated Sentry.io - resolves #90
* Added required NuGet packages.
* Changed target framework from 4.6 to 4.7.2.
* Unhandled exceptions are automatically reported to Sentry.
* A GUID is used to distinguish the sources of the incoming Sentry reports.
* Can be enabled/disabled from settings dialog.

See commit messages for details.

### CODE REVIEW

File: xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs [hunk 1]
```diff
@@ -65,14 +66,18 @@ internal void RegisterEvents()
...
+                if (dialog.ShowDialog() == DialogResult.OK)
                 {
+                    // refresh/update settings dependent functions
                     RefreshModList();
-                    ShowQuickLaunchArgsBasedOnSettings();
                     showHiddenModsToolStripMenuItem.Checked = Settings.ShowHiddenElements;
                     UpdateQuickArgumentsMenu();
+
+                    if (dialog.IsRestartRequired) {
+                        MessageBox.Show("Some changes won't take effect, until after the application has been restarted.", "Info", MessageBoxButtons.OK, MessageBoxIcon.Information);
```
Comment (Xymanek) [inline xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs]: Can't we leave a more persistent (but small) notification somewhere?

File: xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.Designer.cs [hunk 2]
```diff
@@ -232,9 +225,23 @@ private void InitializeComponent()
...
+			// 
+			this.useSentry.AutoSize = true;
+			this.useSentry.Location = new System.Drawing.Point(429, 29);
+			this.useSentry.Name = "useSentry";
+			this.useSentry.Padding = new System.Windows.Forms.Padding(3, 3, 0, 0);
+			this.useSentry.Size = new System.Drawing.Size(170, 20);
+			this.useSentry.TabIndex = 17;
+			this.useSentry.Text = "Send anonymous error reports";
+			this.toolTip.SetToolTip(this.useSentry, "If enabled, unhandled exceptions or other \r\npotential issues are automatically re" +
+        "ported to\r\nour X2CommunityCore Sentry.io account.");
```
Comment (Xymanek) [inline xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.Designer.cs]: "unhandled exceptions" is a bit too technical IMO, I think this phrase needs to be simpler to understand for common players

File: xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs [hunk 3]
```diff
@@ -119,13 +122,15 @@ private void bOK_Click(object sender, EventArgs e)
             Settings.NeverImportTags = neverAdoptTagsAndCatFromprofile.Checked;
             Settings.ShowQuickLaunchArguments = ShowQuickLaunchArgumentsToggle.Checked;
             Settings.CheckForPreReleaseUpdates = checkForPreReleaseUpdates.Checked;
+            Properties.Settings.Default.IsSentryEnabled = useSentry.Checked;
```
Comment (Xymanek) [inline xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs]: Where is this checked/read?

File: xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs (hunk 1 above)
Comment (RevZero) [reply xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs]: Maybe show a label saying "Application restart pending" (probably with an additional tooltip) on the main form?

File: xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs (hunk 3 above)
Comment (RevZero) [reply xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs]: On application start in Program.InitSentry().

File: xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.Designer.cs (hunk 2 above)
Comment (RevZero) [reply xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.Designer.cs]: Yeah, "unexpected errors" would probably be better.

File: xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs (hunk 1 above)
Comment (Xymanek) [reply xcom2-launcher/xcom2-launcher/Forms/MainForm.Events.cs]: 👍

File: xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs (hunk 3 above)
Comment (Xymanek) [reply xcom2-launcher/xcom2-launcher/Forms/SettingsDialog.cs]: Lol, it's first line -,-

**Comment by Xymanek** [comment]
Could you post a screenshot of how the new stuff looks?

File: xcom2-launcher/xcom2-launcher/Program.cs [hunk 4]
```diff
@@ -16,67 +18,173 @@ namespace XCOM2Launcher
...
+            IDisposable sentrySdkInstance = null;
+
+            try
             {
-                CheckForUpdate();
-            }
-#endif
+                sentrySdkInstance = SentrySdk.Init(o =>
+                {
+                    o.Dsn = new Dsn("https://3864ad83bed947a2bc16d88602ac0d87@sentry.io/1478084");
```
Comment (Xymanek) [inline xcom2-launcher/xcom2-launcher/Program.cs]: Maybe make this configurable? ie. not hardcoded

**Comment by Xymanek** [comment]
We should also set this up: https://docs.sentry.io/workflow/releases/?platform=csharp#setting-up-releases

**Comment by Xymanek** [comment]
I think we should enable sentry by default (to handle cases like #55 where the launcher doesn't have chance to show GUI yet) and then at first opportunity ask the user if they want opt-out of it

**Comment by RevZero** [comment]
@Xymanek I would prefer not to enable Sentry and send any data, before the user actually agreed to it. Its cleaner and its practically the first thing that happens. So if the user chooses "yes", we won't miss anything.

**Comment by Xymanek** [comment]
I still think that it would be better (for us) to make it opt-out at first opportunity, but I get your concerns.

I guess we can leave it as-is now and revisit later if we see that we are missing reports

**Pull request merged**